src/fromager/models.py

```diff
@@ -0,0 +1,244 @@
+import logging
+import os
+import pathlib
+import string
+import typing
+
+import yaml
+from packaging.utils import NormalizedName, canonicalize_name
+from packaging.version import Version as BaseVersion
+from pydantic import (
+    BaseModel,
+    BeforeValidator,
+    ConfigDict,
+    Field,
+    GetCoreSchemaHandler,
+    HttpUrl,
+    ValidationInfo,
+    field_validator,
+)
+from pydantic_core import CoreSchema, core_schema
+
+logger = logging.getLogger(__name__)
+
+
+# build directory
+def _before_builddirectory(p: str) -> pathlib.Path:
+    result = pathlib.Path(p)
+    if os.pardir in str(result) or result.is_absolute():
+        raise ValueError("{result!r} is not a relative path")
+    return result
+
+
+BuildDirectory = typing.Annotated[
+    pathlib.Path,
+    BeforeValidator(_before_builddirectory),
+]
+
+
+# version
+class Version(BaseVersion):
+    """Pydantic-aware package version"""
+
+    @classmethod
+    def validate(cls, v: typing.Any, info: core_schema.ValidationInfo) -> BaseVersion:
+        if isinstance(v, BaseVersion):
+            return v
+        return BaseVersion(v)
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls, source_type: typing.Any, handler: GetCoreSchemaHandler
+    ) -> CoreSchema:
+        return core_schema.with_info_plain_validator_function(
+            cls.validate,
+            serialization=core_schema.plain_serializer_function_ser_schema(
+                str, when_used="json"
+            ),
+        )
+
+
+# Changelog
+Changelog = dict[Version, list[str]]
+
+# environment variables map
+EnvVars = dict[str, str | int | float | bool]
```

Comment:
Environment variables are always actually strings, right?

Reply:
Yes, but the idea is that we accept int, float, and bool values in YAML files and convert them to strings. I have now switch to a custom str type with a before validator. The validator converts int, str, bool to str. It also checks for `$(` in str.

Reply:
OK, that makes sense.